java-checks/src/main/java/org/sonar/java/checks/PreparedStatementLoopInvariantCheck.java

```diff
@@ -0,0 +1,210 @@
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    var invocvationCollector = new MethodInvocationCollector(matchers);
+    tree.accept(invocvationCollector);
+
+    var candidatesByLoop = new HashMap<StatementTree, List<Candidate>>();
+    invocvationCollector.invocations.stream()
+      .map(PreparedStatementLoopInvariantCheck::getCandidate)
+      .filter(Objects::nonNull)
+      .forEach(candidate -> {
+        var candidates = candidatesByLoop.computeIfAbsent(candidate.enclosingLoop, key -> new ArrayList<>());
+        candidates.add(candidate);
+      });
+
+    candidatesByLoop.forEach((loop, candidates) -> {
+      var localsCollector = new DeclaredOrAssignedLocalsCollector();
+      loop.accept(localsCollector);
+      candidates.forEach(it -> reportIfLoopInvariant(localsCollector.declaredOrAssignedLocals, it));
+    });
+  }
+
+  private void reportIfLoopInvariant(Set<String> declaredOrAssignedLocals, Candidate candidate) {
+    if (isLoopInvariant(declaredOrAssignedLocals, candidate)) {
+      var secondaryLocation = new JavaFileScannerContext.Location(
+        "Enclosing loop",
+        Objects.requireNonNull(candidate.enclosingLoop)
+      );
+      reportIssue(
+        candidate.invocation,
+        "Move this loop-invariant setter invocation out of this loop.",
+        List.of(secondaryLocation),
+        null
+      );
+    }
+  }
+
+  private static boolean isLoopInvariant(Set<String> declaredOrAssignedLocals, Candidate candidate) {
+    return (candidate.identifierArguments.stream().noneMatch(declaredOrAssignedLocals::contains));
+  }
+
+  private static Candidate getCandidate(MethodInvocationTree invocation) {
+    var identifierArguments = new ArrayList<String>();
+    for (var arg : invocation.arguments()) {
+      if (arg.is(Tree.Kind.IDENTIFIER)) {
+        identifierArguments.add(((IdentifierTree) arg).name());
+      } else {
+        var argValue = ExpressionUtils.resolveAsConstant(arg);
+        if (argValue == null) return null;
+      }
+    }
+
+    var loop = findEnclosingLoop(invocation);
+    if (loop == null) {
+      return null;
+    }
+    return new Candidate(invocation, identifierArguments, loop);
+  }
+
+  private static StatementTree findEnclosingLoop(Tree tree) {
+    while (!Objects.requireNonNull(tree).is(Tree.Kind.METHOD)) {
+      if (isLoop(tree)) {
+        return (StatementTree) tree;
+      }
+      tree = tree.parent();
+    }
+    return null;
+  }
+
+  private static boolean isLoop(Tree tree) {
+    switch (tree.kind()) {
+      case FOR_STATEMENT:
+      case FOR_EACH_STATEMENT:
+      case WHILE_STATEMENT:
+      case DO_STATEMENT:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private static class Candidate {
+    public final MethodInvocationTree invocation;
+    public final List<String> identifierArguments;
+    public final StatementTree enclosingLoop;
```

Comment:
I think that `public` fields are better in such situations and I always use this pattern, but happy to discuss. My reasoning behind this:

- Situation: you have a `private` inner helper class with fields that you want to access in the outer class
- Declaring not only the helper class, but also its fields `private` does now protect the fields twice: once by the `private` of the class, once by the `private` of the field. That's unnecessary and I regard it as a form of defensive programming (in the same way as unnecessary null checks are, just to make things twice as sure)
- While this is already a bad practice, consider you would be later moving the inner class out, because other classes want to use this helper as well. Not only would you need to change the `private` modifier of the class, but also that of every field, and need to rethink which fields can stay private.

I therefore consider it bad practice to declare fields of a private inner class also `private` just because it's possible. What I do:
- For a `private` inner class: a field should be `private` only if the outer class doesn't access it
- For a `public` inner class: a field could still be `private`, depending on whether it belongs to the API of that class or is an implementation detail